FEATURE: adding profile param to s3_options function

ducks opened the pull request:
## Summary

Adds IAM role assumption support for S3 operations, allowing Discourse to use
temporary credentials instead of long-lived static access keys.

## The Problem

The Ruby AWS SDK credential chain has a precedence issue: environment variables
(position 7) take priority over profile-based role assumption (position 10).
Even with a properly configured `~/.aws/config` using `credential_source =
Environment`, the SDK finds `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY` first
and uses them directly, never processing the role assumption.

Upstream declined to fix this for backwards compatibility reasons.

## The Solution

Bypass the broken credential chain by using explicit `Aws::AssumeRoleCredentials`
with an explicit STS client. New env vars control the behavior:

- `DISCOURSE_S3_ROLE_ARN` - IAM role to assume
- `DISCOURSE_S3_ROLE_SESSION_NAME` - Session identifier for CloudTrail auditing

When both are set along with access keys, Discourse creates an STS client with
the base credentials and assumes the specified role, using the temporary
credentials for all S3 operations.

## Credential Flow

1. If `s3_role_session_name` + `s3_role_arn` + keys are set:
   - Create STS client with explicit keys
   - Use AssumeRoleCredentials to get temporary credentials
2. Else if just keys are set:
   - Use keys directly (existing behavior)
3. Else:
   - Let AWS SDK auto-discover (instance profile, ECS task role, etc.)

## Changes

- Add `s3_role_arn` and `s3_role_session_name` settings (SiteSetting + GlobalSetting)
- Remove deprecated `s3_use_iam_profile` setting
- Update S3Helper to use AssumeRoleCredentials when role assumption is configured
- Update FileStore to check correct config source (SiteSetting vs GlobalSetting)
- Simplify S3 config validation (no longer requires credentials if bucket present)

## Configuration

```yaml
# Container env vars
DISCOURSE_S3_ACCESS_KEY_ID: "AKIA..."
DISCOURSE_S3_SECRET_ACCESS_KEY: "..."
DISCOURSE_S3_ROLE_ARN: "arn:aws:iam::123456789012:role/my-site-file-upload"
DISCOURSE_S3_ROLE_SESSION_NAME: "my-site"
```

Or via site settings for UI-configured S3.

## Migration

Existing deployments continue working without changes. Role assumption is
opt-in via the new env vars.

ducks commented on lib/s3_helper.rb:
```diff
@@ -268,14 +269,18 @@ def object(path)
+      raise ArgumentError, "Profile required when using IAM profiles" if profile.blank?
+      opts[:profile] = profile
+    else
+      # Legacy behavior for self-hosters
```
will remove. have just been thinking of legacy behavior as mostly self hosters since we control the future for hosted.

ducks commented on lib/s3_helper.rb:
```diff
@@ -268,14 +269,17 @@ def object(path)
 
-    unless obj.s3_use_iam_profile
+    if obj.s3_use_iam_profile
+      opts[:profile] = profile if profile.present?
```
are you thinking something like this:

```
def self.s3_options(
  region:,
  endpoint: nil,
  http_continue_timeout:,
  use_dualstack_endpoint:,
  use_iam_profile: false,
  file_uploads_profile: nil,
  access_key_id: nil,
  secret_access_key: nil
)
  opts = { region: region }
  opts[:endpoint] = endpoint if endpoint.present?
  opts[:http_continue_timeout] = http_continue_timeout
  opts[:use_dualstack_endpoint] = use_dualstack_endpoint
  
  if use_iam_profile
    opts[:profile] = file_uploads_profile if file_uploads_profile.present?
  else
    opts[:access_key_id] = access_key_id
    opts[:secret_access_key] = secret_access_key
  end
  
  opts
end
```

if so, that feels a little too big for this PR, but I could create a todo to come back after this.

I do agree though that I don't need to pass `profile` and am working on an update to push.

Supermathie commented on lib/site_settings/validations.rb:
```diff
@@ -189,7 +189,9 @@ def validate_backup_location(new_val)
     end
 
-    unless SiteSetting.s3_use_iam_profile
+    # Only validate credentials if user is providing them
```
we should be careful about saying e.g. "user" and this should say "admin" (or whatever we use elsewhere)

blake-discourse commented on app/models/global_setting.rb:
```diff
@@ -99,8 +99,7 @@ def self.use_s3?
         begin
-          if s3_bucket && s3_region &&
-               (s3_use_iam_profile || (s3_access_key_id && s3_secret_access_key))
+          if s3_bucket.present? && s3_region.present?
```
If someone sets s3_bucket and s3_region in their config but doesn’t have any credentials configured (no IAM role, no env vars, no AWS config file), GlobalSetting.use_s3? will return true, and then the actual S3 operations will fail at runtime with credential errors.

Seems like we should check that some creds exist somewhere?

blake-discourse reviewed (dismissed):
Left a few possible suggestions to take a second look at.

ducks committed:
FEATURE: Add IAM role assumption for S3 operations

Adds support for assuming IAM roles for S3 operations using
temporary credentials instead of long-lived static access keys.

New settings:
- s3_role_arn: IAM role to assume
- s3_role_session_name: session identifier for CloudTrail

Credential resolution order:
1. If s3_use_iam_profile is true (deprecated): SDK auto-discovery
2. If role_arn + keys: AssumeRoleCredentials via STS
3. If just keys: static credentials
4. Otherwise: SDK auto-discovery (instance profile, ECS task role)

Deprecates s3_use_iam_profile with removal planned for 3.6.
Sites with it enabled continue working with a deprecation warning.

Supermathie reviewed (changes requested):
I meant to request changes, not approve

Supermathie commented on config/site_settings.yml:
```diff
@@ -2226,6 +2224,12 @@ files:
   s3_secret_access_key:
     default: ""
     secret: true
+  s3_role_session_name:
```
Neither this nor `s3_role_arn` needs to be secret.

Of course, neither does `s3_access_key_id`.

Supermathie reviewed (changes requested):
Other than settings being secret when they don't need to be, LGTM

tgxworld commented:
> Even with a properly configured `~/.aws/config` using `credential_source = Environment`, the SDK finds `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY` first
> and uses them directly, never processing the role assumption.

I'm not sure I follow the problem statement here. Discourse doesn't set either `AWS_ACCESS_KEY_ID` or `AWS_SECRET_ACCESS_KEY` directly in our env so why is this a problem?

tgxworld commented:
> * Remove deprecated `s3_use_iam_profile` setting

This is also not quite true. The site setting has never been deprecated from what I can tell and we are in fact removing it without deprecation. What happens if a site has s3 keys configured but also `s3_use_iam_profile` set to `true`? Removing this site setting without warning will cause the s3 keys to be used instead.

tgxworld reviewed (changes requested):
There are pending security and backwards compatibility issues to resolve.

tgxworld commented on lib/aws_credentials.rb:
```diff
@@ -0,0 +1,172 @@
+
+# Credentials from GlobalSetting (ENV vars)
+# Takes precedence over SiteAwsCredentials
+class GlobalAwsCredentials < AwsCredentials
```
Let's move this to its own file as we don't usually mix multiple classes into a single file.

tgxworld commented on lib/aws_credentials.rb:
```diff
@@ -0,0 +1,172 @@
+  end
+
+  def configured?
+    source.s3_bucket.present? && source.s3_region.present?
```
I don't quite follow the logic here. Why is `s3_bucket` and `s3_region` used to determine if `GlobalAwsCredentials ` is configured? We are dealing with credentials here so shouldn't the keys used for credentials be used to determine if we should use  `GlobalAwsCredentials` or not?

tgxworld commented on config/site_settings.yml:
```diff
@@ -2219,11 +2219,8 @@ files:
-    default: false
   s3_access_key_id:
     default: ""
-    secret: true
```
What was the rationale in removing this in the review feedback?

ducks commented:
> > > > Even with a properly configured `~/.aws/config` using `credential_source = Environment`, the SDK finds `AWS_ACCESS_KEY_ID`/`AWS_SECRET_ACCESS_KEY` first
> > > > and uses them directly, never processing the role assumption.
> > > 
> > > 
> > > I'm not sure I follow the problem statement here. Discourse doesn't set either `AWS_ACCESS_KEY_ID` or `AWS_SECRET_ACCESS_KEY` directly in our env so why is this a problem?
> > 
> > 
> > to use the credential chain, you would need to set those and that's where the problem comes in. if you do set those, they are always used as direct credentials instead of checking the `credential_source` config option.
> 
> Are we planning to set `AWS_ACCESS_KEY_ID ` or `AWS_SECRET_ACCESS_KEY` somewhere?

yes, these will be set in the container to be used as credentials to assume a role instead of used directly as they are now.

discourse-patch-triage[bot] committed:
Merge remote-tracking branch 'origin/main' into feature/assuming-iam-role-s3-pt2

# Conflicts:
#	Gemfile.lock
#	app/services/video_conversion/aws_media_convert_adapter.rb
#	config/discourse_defaults.conf
#	config/site_settings.yml
#	lib/s3_helper.rb
#	spec/services/video_conversion/aws_media_convert_adapter_spec.rb
